### PR TITLE
nautilus: bluestore: Add protection against bluefs log file growth

### PR DESCRIPTION
**Pull request by aclamk**
This fixes https://tracker.ceph.com/issues/46194.

### CODE REVIEW

**Comment by smithfarm** [comment]
@aclamk  The commit messages don't have any "Conflicts" sections, which indicates that there were no conflicts when you did `git cherry-pick -x`. Is that really the case, though?

**Comment by smithfarm** [comment]
(I took the liberty of modifying the PR title... Since backport PR titles become lines in the release notes, we try to keep them uniform.)

**Comment by aclamk** [comment]
@smithfarm My bad. Sorry. There were conflicts in 2 of them. Let me redo cherry-picking.

**Comment by aclamk** [comment]
@smithfarm Done. Is ok?

**Comment by smithfarm** [comment]
> @smithfarm Done. Is ok?

I'm fine with it :-)

It looks like this is a backport of https://github.com/ceph/ceph/pull/35473 to nautilus. That master PR contains 4 commits, but this nautilus backport only three. Maybe you could mention that you omitted one of the commits (and explain the reason) in the PR description? I'm thinking that might make life easier for the reviewers and others who might trying to understand this PR.

**Comment by aclamk** [comment]
@smithfarm I am thinking that adding test that is not used is easier then explanations.

File: src/test/objectstore/test_bluefs.cc [hunk 1]
```diff
@@ -620,6 +651,45 @@ TEST(BlueFS, test_replay) {
   rm_temp_bdev(fn);
 }
 
+TEST(BlueFS, test_replay_growth) {
+  uint64_t size = 1048576LL * (2 * 1024 + 128);
+  TempBdev bdev{size};
```
Comment (neha-ojha) [inline src/test/objectstore/test_bluefs.cc:656]: This is causing a make check failure https://jenkins.ceph.com/job/ceph-pull-requests/63024/consoleFull#-15327964252a811ea2-3e7b-466b-84b4-d13df7e35809

**Comment by neha-ojha** [comment]
@smithfarm I am removing needs-qa until the make check failure is fixed.

**Comment by aclamk** [comment]
jenkins make check please

**Comment by aclamk** [comment]
jenkins retest this please

File: src/os/bluestore/BlueFS.cc [hunk 2]
```diff
@@ -1556,7 +1556,10 @@ int BlueFS::_read_random(
       s_lock.unlock();
       uint64_t x_off = 0;
       auto p = h->file->fnode.seek(off, &x_off);
-      uint64_t l = std::min(p->length - x_off, static_cast<uint64_t>(len));
+      ceph_assert(p != h->file->fnode.extents.end());
+      uint64_t l = std::min(p->length - x_off, len);
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc:1560]: actually this has got modifications from multiple PRs. Original one you're backporting includes just ceph_assert()

**Comment by smithfarm** [comment]
Marking DNM pending @aclamk 's response to @ifed01 's question.

**Comment by smithfarm** [comment]
@ifed01 @neha-ojha Ping - it looks like @aclamk force-pushed this since you left your comments. Could you have another look?

**Review by ifed01 — APPROVED** [review]

**Review by neha-ojha — APPROVED** [review]

**Comment by yuriw** [comment]
[wip-yuri6-testing-2021-01-28-1020-nautilus](https://trello.com/c/OELtEJLp/1129-wip-yuri6-testing-2021-01-28-1020-nautilus)

**Pull request merged**